### PR TITLE
MSDKUI-1551 [Android] RP - Canceling setting waypoint creates an empt…

### PR DESCRIPTION
**Pull request by tstaszak89**
…y waypoint item to list and results in a crash

Route Planner - empty waypoints will no longer be added when exiting waypoint
selection using X icon. Updated unit tests and documentation.

Signed-off-by: Tadeusz Staszak <36914750+tstaszak89@users.noreply.github.com>

### CODE REVIEW

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt [hunk 1]
```diff
@@ -105,12 +107,17 @@ class WaypointSelectionFragment() : RetainFragment(),
      * Update fragment with given [GeoCoordinate]. This will do reverse geo-coding to get name for
      * the given [GeoCoordinate].
      *
-     * @param cord  [GeoCoordinate]
+     * @param cord
+     *         [GeoCoordinate]
      */
     fun updateCord(cord: GeoCoordinate) {
```
Comment (datasun) [inline MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt]: What is updated here? The new GeoCoordinate to be associated with this Waypoint?

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt [hunk 2]
```diff
@@ -149,9 +156,22 @@ class WaypointSelectionFragment() : RetainFragment(),
     interface Listener {
 
         /**
-         * Callback to be called when user select a waypoint.
+         * Callback to be called when a user cancelled waypoint selection.
+         *
+         * @param index
+         *         an integer if you want to track waypoint entry, useful like using in list, null otherwise.
```
Comment (datasun) [inline MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt]: the index of the waypoint entry that was unselected?

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt [hunk 3]
```diff
@@ -149,9 +156,22 @@ class WaypointSelectionFragment() : RetainFragment(),
...
+         *         WaypointEntry object.
+         */
+        fun onWaypointSelectionCancelled(index: Int?, current: WaypointEntry?)
+
+        /**
+         * Callback to be called when a user has selected a waypoint.
          *
-         * @param index a integer if you want to track waypoint entry, useful like using in list, null otherwise.
+         * @param index
+         *         an integer if you want to track waypoint entry, useful like using in list, null otherwise.
```
Comment (datasun) [inline MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt]: same here

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionPresenter.kt [hunk 4]
```diff
@@ -92,8 +97,10 @@ class WaypointSelectionPresenter() : BasePresenter<CommonContracts.WaypointSelec
     /**
      * Updates position with given [WaypointEntry].
      *
-     * @param index index of entry to keep track of entry.
-     * @param entry [WaypointEntry]
+     * @param index
+     *         index of entry to keep track of entry.
```
Comment (datasun) [inline MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionPresenter.kt]: the new position of the waypoint entry?

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionPresenter.kt [hunk 5]
```diff
@@ -104,7 +111,8 @@ class WaypointSelectionPresenter() : BasePresenter<CommonContracts.WaypointSelec
      * Update fragment with given [GeoCoordinate]. This will do reverse geo-coding to get name for
      * the given [GeoCoordinate].
      *
-     * @param cord [GeoCoordinate]
+     * @param cord
+     *         [GeoCoordinate]
      */
     fun updateCord(cord: GeoCoordinate) {
```
Comment (datasun) [inline MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionPresenter.kt]: same here

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt [hunk 6]
```diff
@@ -185,14 +187,38 @@ class RoutePlannerFragment() : RetainFragment(), RoutingContracts.RoutePlanner {
...
-     * @param updateBack if back button needs to be updated.
-     * @param updateTitle if title needs to be updated.
-     * @param updateRightIcon if right icon needs to be update.
+     * @param updateBack
+     *         if back button needs to be updated.
+     * @param updateTitle
+     *         if title needs to be updated.
+     * @param updateRightIcon
+     *         if right icon needs to be update.
      */
```
Comment (datasun) [inline MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt]: updated

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt [hunk 7]
```diff
@@ -185,14 +187,38 @@ class RoutePlannerFragment() : RetainFragment(), RoutingContracts.RoutePlanner {
...
     fun updateActionBar(updateBack: Boolean = true, updateTitle: Boolean = true, updateRightIcon: Boolean = true) {
         presenter.updateActionBar(updateBack = updateBack, updateTitle = updateTitle, updateRightIcon = updateRightIcon)
     }
 
+    /**
+     * Make action on waypoint selection cancelled - if this waypoint is invalid then remove it from list.
```
Comment (datasun) [inline MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt]: Handle an unselected waypoint. If the waypoint is invalid, then it will be removed from the waypoint list.

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt [hunk 8]
```diff
@@ -185,14 +187,38 @@ class RoutePlannerFragment() : RetainFragment(), RoutingContracts.RoutePlanner {
...
+     *
+     * @param index
+     *         index within the list where selection been cancelled.
+     * @param current
+     *         [WaypointEntry] that selection been cancelled.
```
Comment (datasun) [inline MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt]: the unselected [WaypointEntry].

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt [hunk 9]
```diff
@@ -185,14 +187,38 @@ class RoutePlannerFragment() : RetainFragment(), RoutingContracts.RoutePlanner {
...
      */
     fun updateActionBar(updateBack: Boolean = true, updateTitle: Boolean = true, updateRightIcon: Boolean = true) {
         presenter.updateActionBar(updateBack = updateBack, updateTitle = updateTitle, updateRightIcon = updateRightIcon)
     }
 
+    /**
+     * Make action on waypoint selection cancelled - if this waypoint is invalid then remove it from list.
+     *
+     * @param index
+     *         index within the list where selection been cancelled.
```
Comment (datasun) [inline MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt]: the index of the unselected [WaypointEntry].

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt (hunk 1 above)
Comment (datasun) [inline MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt]: Maybe this method can be renamed in a separate commit?

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt [hunk 10]
```diff
@@ -105,12 +107,17 @@ class WaypointSelectionFragment() : RetainFragment(),
      * Update fragment with given [GeoCoordinate]. This will do reverse geo-coding to get name for
      * the given [GeoCoordinate].
      *
-     * @param cord  [GeoCoordinate]
+     * @param cord
```
[the remark this answers is not in the document]
Comment (tstaszak89) [reply MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt]: I will rename it into coordinate

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt (hunk 1 above)
Comment (tstaszak89) [reply MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt]: Yes, it updates coordinate of current waypoint and start reverse geocoding request.

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt (hunk 2 above)
Comment (tstaszak89) [reply MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt]: I would suggest: the index of the waypoint entry on route planner waypoints list.

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt (hunk 3 above)
Comment (tstaszak89) [reply MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionFragment.kt]: I would suggest: the index of the waypoint entry on route planner waypoints list.

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionPresenter.kt (hunk 4 above)
Comment (tstaszak89) [reply MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionPresenter.kt]: I agree, update comment.

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionPresenter.kt (hunk 5 above)
Comment (tstaszak89) [reply MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/common/mapselection/WaypointSelectionPresenter.kt]: Update to coordinate

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt (hunk 6 above)
Comment (tstaszak89) [reply MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt]: I agree, I will update comment.

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt (hunk 7 above)
Comment (tstaszak89) [reply MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt]: I agree, I will update comment.

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt (hunk 8 above)
Comment (tstaszak89) [reply MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt]: I agree, I will update comment.

File: MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt (hunk 9 above)
Comment (tstaszak89) [reply MSDKUIDemo/MSDKUIApp/src/main/java/com/here/msdkuiapp/routing/RoutePlannerFragment.kt]: I agree, I will update comment.
